### PR TITLE
Prometheus: Check for errors on json response parsing

### LINKED ISSUE #73747: Prometheus: No error returned when response_limit is hit with /ds/query
When the Grafana configuration `response_limit` is used and hit, the prometheus datasource does not return a warning or an error, but instead no data, empty items, or partial data with no indication of a limited response.

This seems to be because we are not checking `iter.Error` through `pkg/util/converter`.

https://github.com/grafana/grafana/blob/d293b08e525c6b30f7d10bd6c589680bcb705793/pkg/util/converter/prom.go#L28

https://github.com/grafana/grafana/blob/d293b08e525c6b30f7d10bd6c589680bcb705793/pkg/infra/httpclient/httpclientprovider/response_limit_middleware.go#L13

Examples:

<details>
    <summary>With a setting of `10` bytes, I get a response that looks like "No Data"</summary>

```json
{
    "results": {
        "A": {
            "status": 200,
            "frames": [
                {
                    "schema": {
                        "refId": "A",
                        "meta": {
                            "typeVersion": [
                                0,
                                0
                            ],
                            "executedQueryString": "Expr: up\nStep: 15s"
                        },
                        "fields": []
                    },
                    "data": {
                        "values": []
                    }
                }
            ]
        }
    }
}
```

</details>

<details>
    <summary>If I increase it 100 bytes, get an "empty response"</summary>

```json
{
    "results": {
        "A": {
            "status": 200,
            "frames": [
                {
                    "schema": {
                        "refId": "A",
                        "meta": {
                            "type": "timeseries-multi",
                            "typeVersion": [
                                0,
                                1
                            ],
                            "custom": {
                                "resultType": "matrix"
                            },
                            "executedQueryString": "Expr: up\nStep: 15s"
                        },
                        "fields": [
                            {
                                "name": "Time",
                                "type": "time",
                                "typeInfo": {
                                    "frame": "time.Time"
                                },
                                "config": {
                                    "interval": 15000
                                }
                            },
                            {
                                "name": "up",
                                "type": "number",
                                "typeInfo": {
                                    "frame": "float64"
                                },
                                "labels": {
                                    "__name__": "up",
                                    "instance": ""
                                }
                            }
                        ]
                    },
                    "data": {
                        "values": [
                            [],
                            []
                        ]
                    }
                }
            ]
        }
    }
}
```

</details>

And larger numbers  (e.g. 300 bytes) can result in partial data.

With a short temp test in util/converter and some spew, verified that we will see the error in `iter.Error`:

```
func TestReadLimited(t *testing.T) {
	f, err := os.Open(path.Join("testdata", "prom-matrix"+".json"))
	require.NoError(t, err)
	b := httpclient.MaxBytesReader(f, 90)
	iter := jsoniter.Parse(jsoniter.ConfigDefault, b, 1024)
	rsp := ReadPrometheusStyleResult(iter, Options{})
}
```

Then in one of the `for l1Field := iter.ReadObject(); l1Field != ""; l1Field = iter.ReadObject()` blocks `iter.Error` did get set to `(*fmt.wrapError)(0xc000408960)(error: http: response body too large, response limit is set to: 90)`. So we need to go and check errors on the calls in the parser I believe.

### PR DESCRIPTION
**Pull request by kylebrandt**
**What is this feature?**

Checking errors in the jsoniter response parser. Currently if a response is limited by the `response_limit` setting, it silent fails -- see issue.

This checks for errors (including limit, but also if there is malformed json) and returns the error.

image

Fixes https://github.com/grafana/grafana/issues/73747

**Special notes for your reviewer:**

* The JSON dataframe parser in the plugin sdk may need this as well - haven't checked.

* Maybe there is a reason that I am not seeing, but really wishing the jsoniter `Iterator` methods returned errors instead of setting `iter.Error`. **Edit**: See this (incomplete) alternative https://github.com/grafana/grafana/pull/73861) 

**Impact**:
Assuming I didn't screw anything up, I can imagine the following impact:
 - If something was sending malformed JSON, the previous code might have accepted that and partially processed the data
 - Before there may have been partial data with no error, where as no there may be errors but they will have no partial results, just the error.

So in short this may surface some problems that were being masked before. This code is used by both the prometheus and loki datasources.

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: pkg/util/converter/prom.go [hunk 1]
```diff
@@ -243,20 +382,41 @@ func readLabelsOrExemplars(iter *jsoniter.Iterator) (*data.Frame, [][2]string) {
...
+						}
+						ts := timeFromFloat(f)
 						timeField.Append(ts)
 
 					case "labels":
 						max := 0
-						for _, pair := range readLabelsAsPairs(iter) {
+						pairs, err := readLabelsAsPairs(iter)
+						if err != nil {
+							return nil, nil, iter.Error
```
Comment (kylebrandt) [inline pkg/util/converter/prom.go:417]: should return `err` instead of `iter.Error` here.

File: pkg/util/converter/prom.go [hunk 2]
```diff
@@ -69,14 +107,26 @@ func ReadPrometheusStyleResult(iter *jsoniter.Iterator, opt Options) backend.Dat
 	return rsp
```
Comment (ryantxu) [inline pkg/util/converter/prom.go:107]:
```suggestion
    if iter.Error != nil && rsp.Error == nil {
			rsp.Error = iter.Error
	}
	return rsp
```
I wonder if the explicit checks are needed within each element, or just checking once at the end 🤦🏻

Comment (kylebrandt) [reply pkg/util/converter/prom.go:107]: I think it is best to stop as soon as the error as occurred, or is harder to predicate the behavior and later errors may mask previous errors.

**Comment by kylebrandt** [comment]
I think we want to wait until we have FE support to merge this, as with ML, New Warnings (for join/union drops), Pending Error handling change, and this -- we are starting to build up an in-balance in inventory of BE code in main for SSE vs FE.

**Comment by kylebrandt** [comment]
Another option to consider that we could swap this out for... something like:

```go
package jsonitere

import j "github.com/json-iterator/go"

type Iterator struct {
	i j.Iterator
}

func (iter *Iterator) CurrentBuffer() string {
	return iter.i.CurrentBuffer()
}

func (iter *Iterator) ReadAny() (j.Any, error) {
	return iter.i.ReadAny(), iter.i.Error
}

// etc...
```

File: pkg/util/converter/jsonitere/jsonitere.go [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+func NewIterator(i *j.Iterator) *Iterator {
+	return &Iterator{i}
+}
+
+func (iter *Iterator) Read() (interface{}, error) {
+	return iter.i.Read(), iter.i.Error
+}
+
+func (iter *Iterator) ReadAny() (j.Any, error) {
+	// Clear iter.i.Error?
```
Comment (itsmylife) [inline pkg/util/converter/jsonitere/jsonitere.go:22]: This is something will be done later?

**Review by itsmylife — APPROVED** [review]
I tested this on explore and a couple of dashboards. Everything is working. 
So LGTM!

File: pkg/util/converter/jsonitere/jsonitere.go (hunk 3 above)
Comment (kylebrandt) [reply pkg/util/converter/jsonitere/jsonitere.go:22]: I'll make a note in the pkg docs.

File: pkg/util/converter/jsonitere/jsonitere.go [hunk 4]
```diff
@@ -0,0 +1,58 @@
+// package jsonitere wraps json-iterator/go's Iterator methods with error returns
+// so linting can catch unchecked errors.
+package jsonitere
```
Comment (kylebrandt) [inline pkg/util/converter/jsonitere/jsonitere.go:3]: I think I may end up moving this into our go-plugin-sdk repo since the dataframe parser uses `json-iterator/go` as well. But it will be easy enough to remove this and change any imports if/when that happens.
